src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/CpsPackageReferenceProject.cs

```diff
@@ -358,10 +358,48 @@ private async Task SetPackagePropertyValueAsync(IProjectProperties metadata, str
                 propertyValue);
         }
 
-        public override async Task<bool> UninstallPackageAsync(PackageIdentity packageIdentity, INuGetProjectContext nuGetProjectContext, CancellationToken token)
+        public override Task<bool> UninstallPackageAsync(PackageIdentity packageIdentity, INuGetProjectContext nuGetProjectContext, CancellationToken token)
         {
-            var configuredProject = await _unconfiguredProject.GetSuggestedConfiguredProjectAsync();
-            await configuredProject?.Services.PackageReferences.RemoveAsync(packageIdentity.Id);
+            throw new InvalidOperationException("For CPSPackageReferenceProject, the uninstall method that can handle conditional targeting must be called instead.");
+        }
+
+        public override async Task<bool> UninstallPackageAsync(string packageId, BuildIntegratedInstallationContext installationContext, CancellationToken token)
+        {
+            token.ThrowIfCancellationRequested();
+            if (string.IsNullOrEmpty(packageId)) throw new ArgumentException(string.Format(Strings.Argument_Cannot_Be_Null_Or_Empty, nameof(packageId)));
+            if (installationContext == null) throw new ArgumentNullException(nameof(installationContext));
+
+            if (installationContext.SuccessfulFrameworks.Any() && installationContext.UnsuccessfulFrameworks.Any())
+            {
+                var conditionalService = _unconfiguredProject
+                    .Services
+                    .ExportProvider
+                    .GetExportedValue<IConditionalPackageReferencesService>();
+
+                if (conditionalService == null)
+                {
+                    throw new InvalidOperationException(string.Format(
+                        CultureInfo.CurrentCulture,
+                        Strings.UnableToGetCPSPackageInstallationService,
+                        ProjectFullPath));
+                }
+
+                foreach (var framework in installationContext.SuccessfulFrameworks)
+                {
+                    string originalFramework;
+                    if (!installationContext.OriginalFrameworks.TryGetValue(framework, out originalFramework))
```

Comment:
My understanding is that the `installationContext.SuccessfulFrameworks` is a subset of `installationContext.OriginalFrameworks`. If this is accurate then I guess the dictionary will always contain the `framework` and we dont need to invoke `framework.GetShortFolderName()` method. Please let me know if I am missing something.

Reply:
I was just being defensive here. I think @zivkan is gonna have to change these APIs when working on aliases, and can make them sctricter. For example, he'd probably only need to pass a string and not a framework as a condition.

Reply:
Digging deeper, probably being defensive here is unnecessary. 

Right now it's the only pending change, so if you'd like me to change it, let me know and I'd be happy to do it.

Reply:
No worries, Nikolche. I am okay with you creating a follow-up PR to address this comment.